jumpscale/packages/tfgrid_solutions/utils.py

```diff
@@ -0,0 +1,92 @@
+import time
+from decimal import Decimal
+from jumpscale.loader import j
+
+
+def get_payment_amount(pool):
+    escrow_info = pool.escrow_information
+    resv_id = pool.reservation_id
+    escrow_address = escrow_info.address
+    escrow_asset = escrow_info.asset
+    total_amount = escrow_info.amount
+    if not total_amount:
+        return
+    total_amount_dec = Decimal(total_amount) / Decimal(1e7)
+    total_amount = "{0:f}".format(total_amount_dec)
+    return resv_id, escrow_address, escrow_asset, total_amount
+
+
+def pay_pool(pool_info):
+    # Get the amount of payment
+    resv_id, escrow_address, escrow_asset, total_amount = get_payment_amount(pool_info)
+    # Get the user wallets
+    wallets = j.sals.reservation_chatflow.reservation_chatflow.list_wallets()
+    for wallet_name, wallet_val in wallets.items():
+        j.logger.info(f"Trying to pay pool {pool_info.pool_id} with wallet {wallet_name}")
+        try:
+            wallet_val.transfer(
+                destination_address=escrow_address, amount=total_amount, asset=escrow_asset, memo_text=f"p-{resv_id}"
+            )
+            return True
+        except:
+            j.logger.warning(f"failed to pay pool {pool_info.pool_id} with wallet {wallet_name}")
+    return False
+
+
+def check_pool_expiration(pool, threshold_days=14):
+    remaining_days = (pool.empty_at - time.time()) / 86400  # converting seconds to days
+    return remaining_days < threshold_days and remaining_days > 0
+
+
+def calculate_pool_units(pool, days=14):
+    cu = pool.active_cu * 60 * 60 * 24 * days
+    su = pool.active_su * 60 * 60 * 24 * days
+    return cu, su
+
+
+def auto_extend_pools():
+    pools = j.sals.zos.get().pools.list()
+    for pool in pools:
+        if check_pool_expiration(pool):
+            cu, su = calculate_pool_units(pool)
+            auto_extend_pool(pool.pool_id, int(cu), int(su))
+
+
+def auto_extend_pool(pool_id, cu, su, currencies=None):
+    currencies = currencies or ["TFT"]
+    try:
+        pool_info = j.sals.zos.get().pools.extend(pool_id, cu, su, currencies=currencies)
+    except Exception as e:
+        send_pool_info_mail(pool_id)
+        j.logger.error(f"Error happend during extending the pool, {str(e)}")
+    if not pay_pool(pool_info):
+        send_pool_info_mail(pool_id)
```

Comment:
do we have some kind of validation on emails or what happens in case sending emails failed?

Reply:
it will send the other emails, but anyway we added mail validations when adding escalations emails
![Screenshot from 2020-10-27 13-42-32](https://user-images.githubusercontent.com/1689456/97297172-4f361600-185a-11eb-8ddb-9dfa52d5df73.png)